onnxruntime/core/providers/cpu/object_detection/non_max_suppression_helper.h

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <stdint.h>
+
+#ifdef __NVCC__
+#include "core/providers/cuda/cu_inc/common.cuh"
+#define ORT_DEVICE __device__
+#define HelperMin(a, b) _Min(a, b)
+#define HelperMax(a, b) _Max(a, b)
+#else
+#include <algorithm>
+#define ORT_DEVICE
+#define HelperMin(a, b) std::min(a, b)
+#define HelperMax(a, b) std::max(a, b)
+#endif
+
+namespace onnxruntime {
+
+struct PrepareContext {
+  const float* boxes_data_ = nullptr;
+  int64_t boxes_size_ = 0ll;
+  const float* scores_data_ = nullptr;
+  int64_t scores_size_ = 0ll;
+  // The below are ptrs since they cab be device specific
+  const int64_t* max_output_boxes_per_class_ = nullptr;
+  const float* score_threshold_ = nullptr;
+  const float* iou_threshold_ = nullptr;
+  int64_t num_batches_ = 0;
+  int64_t num_classes_ = 0;
+  int64_t num_boxes_ = 0;
+};
+
+struct SelectedIndex {
+  ORT_DEVICE
+  SelectedIndex(int64_t batch_index, int64_t class_index, int64_t box_index)
+      : batch_index_(batch_index), class_index_(class_index), box_index_(box_index) {}
+  SelectedIndex() = default;
+  int64_t batch_index_ = 0;
+  int64_t class_index_ = 0;
+  int64_t box_index_ = 0;
+};
+
+#ifdef __NVCC__
+namespace cuda {
+#endif
+namespace nms_helpers {
+
+ORT_DEVICE
+inline void MaxMin(float lhs, float rhs, float& min, float& max) {
+  if (lhs >= rhs) {
+    min = rhs;
+    max = lhs;
+  } else {
+    min = lhs;
+    max = rhs;
+  }
+}
+
+ORT_DEVICE
+inline bool SuppressByIOU(const float* boxes_data, int64_t box_index1, int64_t box_index2,
```

Comment:
>inline bool SuppressByIOU [](start = 0, length = 25)

Does this need to be in a header or could it be in a .cc so we don't impact the binary size as much? I know it's not a huge amount of code but just wondering whether the general pattern should be to avoid inlining unless it's performance critical code.

Reply:
This code was written to be shared between CPU and CUDA kernels. Headers are for sharing code. There is not a way to avoid duplication since this needs to be compiled twice, first for CPU and then inside CUDA kernel with a different signature by a different compiler.

---
In reply to: [298060708](https://github.com/microsoft/onnxruntime/pull/1303#discussion_r298060708) [](ancestors = 298060708)